src/EditorFeatures/Core/Implementation/IntelliSense/Helpers.cs

```diff
@@ -1,27 +1,85 @@
+        internal static IEnumerable<object> BuildClassifiedTextElements(ImmutableArray<TaggedText> taggedTexts, Document document, IEnumerable<Lazy<IStreamingFindUsagesPresenter>> streamingPresenters)
+        {
+            var index = 0;
+            return BuildClassifiedTextElements(taggedTexts, ref index, document, streamingPresenters);
+        }
+
+        private static IReadOnlyCollection<object> BuildClassifiedTextElements(ImmutableArray<TaggedText> taggedTexts, ref int index, Document document, IEnumerable<Lazy<IStreamingFindUsagesPresenter>> streamingPresenters)
         {
             // This method produces a sequence of zero or more paragraphs
             var paragraphs = new List<object>();
 
             // Each paragraph is constructed from one or more lines
-            var currentParagraph = new List<ClassifiedTextElement>();
+            var currentParagraph = new List<object>();
 
             // Each line is constructed from one or more inline elements
             var currentRuns = new List<ClassifiedTextRun>();
 
-            foreach (var part in taggedTexts)
+            for (; index < taggedTexts.Length; index++)
             {
+                var part = taggedTexts[index];
+                if (part.Tag == TextTags.ContainerStart)
+                {
+                    if (currentRuns.Count > 0)
+                    {
+                        // This line break means the end of a line within a paragraph.
+                        currentParagraph.Add(new ClassifiedTextElement(currentRuns));
+                        currentRuns.Clear();
+                    }
+
+                    index++;
+                    var nestedElements = BuildClassifiedTextElements(taggedTexts, ref index, document, streamingPresenters);
+                    if (nestedElements.Count <= 1)
+                    {
+                        currentParagraph.Add(new ContainerElement(
+                            ContainerElementStyle.Wrapped,
+                            new ClassifiedTextElement(new ClassifiedTextRun(ClassificationTypeNames.Text, part.Text)),
+                            new ContainerElement(ContainerElementStyle.Stacked, nestedElements)));
+                    }
+                    else
+                    {
+                        currentParagraph.Add(new ContainerElement(
+                            ContainerElementStyle.Wrapped,
+                            new ClassifiedTextElement(new ClassifiedTextRun(ClassificationTypeNames.Text, part.Text)),
+                            new ContainerElement(
+                                ContainerElementStyle.Stacked,
+                                nestedElements.First(),
+                                new ContainerElement(
+                                    ContainerElementStyle.Stacked | ContainerElementStyle.VerticalPadding,
+                                    nestedElements.Skip(1)))));
+                    }
+
+                    continue;
```

Comment:
could we consider extracting out everything above (into something like StartContainerElement(...))?  I thnk it would make it much easier to understand things if we just had a top level lopo, with the high level decision making, and then have the individual functions that are responsible for each piece of work.  intermingling them makes it harder to keep track and understand what's going on for me.